rancher/charts
# Updating DRIVER_REG_SOCK_PATH for windows daemonset

## Issue: [Unable to complete installation of rancher-vsphere-csi on hybrid cluster](https://github.com/rancher/rke2/issues/4025)
 

## Problem
When attempting to install the rancher-vsphere-csi chart on a hybrid (windows/linux) cluster, the node-driver-registrar pod on the windows nodes remains in a crashLoopBackoff state and cannot register properly. The logs show the following error:
```
I0316 12:22:40.230678 6952 main.go:166] Version: v2.5.1
I0316 12:22:40.231742 6952 main.go:167] Running node-driver-registrar in mode=registration
I0316 12:22:40.231784 6952 main.go:191] Attempting to open a gRPC connection with: "unix://C:\\\\csi\\\\csi.sock"
I0316 12:22:40.231784 6952 connection.go:154] Connecting to unix://C:\\csi\\csi.sock
W0316 12:22:50.249110 6952 connection.go:173] Still connecting to unix://C:\\csi\\csi.sock
I0316 12:22:54.449428 6952 main.go:198] Calling CSI driver to discover driver name
I0316 12:22:54.451098 6952 connection.go:183] GRPC call: /csi.v1.Identity/GetPluginInfo
I0316 12:22:54.451098 6952 connection.go:184] GRPC request: {}
I0316 12:22:54.489559 6952 connection.go:186] GRPC response: {"name":"csi.vsphere.vmware.com","vendor_version":"v2.6.2"}
I0316 12:22:54.489559 6952 connection.go:187] GRPC error: <nil>
I0316 12:22:54.489559 6952 main.go:208] CSI driver name: "csi.vsphere.vmware.com"
I0316 12:22:54.492531 6952 node_register.go:53] Starting Registration Server at: /registration/csi.vsphere.vmware.com-reg.sock
I0316 12:22:54.493901 6952 node_register.go:62] Registration Server started at: /registration/csi.vsphere.vmware.com-reg.sock
I0316 12:22:54.496468 6952 node_register.go:92] Skipping HTTP server because endpoint is set to: ""
I0316 12:22:54.593692 6952 main.go:102] Received GetInfo call: &InfoRequest{}
I0316 12:22:54.595982 6952 main.go:109] "Kubelet registration probe created" path="\\var\\lib\\kubelet\\plugins\\csi.vsphere.vmware.com\\registration"
I0316 12:22:56.947417 6952 main.go:120] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:false,Error:RegisterPlugin error -- plugin registration failed with err: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix \\\\var\\\\lib\\\\kubelet\\\\plugins\\\\csi.vsphere.vmware.com\\\\csi.sock: connect: A socket operation was attempted to an unreachable network.",}
E0316 12:22:56.947417 6952 main.go:122] Registration process failed with error: RegisterPlugin error -- plugin registration failed with err: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix \\\\var\\\\lib\\\\kubelet\\\\plugins\\\\csi.vsphere.vmware.com\\\\csi.sock: connect: A socket operation was attempted to an unreachable network.", restarting registration container. 
```
This is the result of the environment variable being set for DRIVER_REG_SOCK_PATH in the daemonset for the node-driver-registrarr on windows nodes being an invalid windows path:
```
containers:
    - args:
        - '--v=5'
        - '--csi-address=$(ADDRESS)'
        - '--kubelet-registration-path=$(DRIVER_REG_SOCK_PATH)'
      env:
        - name: ADDRESS
          value: unix://C:\\csi\\csi.sock
        - name: DRIVER_REG_SOCK_PATH
          value: \\var\\lib\\kubelet\\plugins\\csi.vsphere.vmware.com\\csi.sock
```

## Solution
I was able to remediate this issue by updating the environment variable in the windows node-driver-registrarr daemonset to point to a valid path on the filesystem:
```
  env:
    - name: ADDRESS
       value: unix://C:\\csi\\csi.sock
    - name: DRIVER_REG_SOCK_PATH
       value: C:\\var\\lib\\kubelet\\plugins\\csi.vsphere.vmware.com\\csi.sock
```

## Testing
As shown below, once this change was make, the daemonset was able to be properly deployed. 
```
I0809 14:37:47.682853    1700 main.go:167] Version: v2.7.0
2023-08-09T14:37:47.686514000+01:00 I0809 14:37:47.682853    1700 main.go:168] Running node-driver-registrar in mode=registration
2023-08-09T14:37:47.686514000+01:00 I0809 14:37:47.686514    1700 main.go:192] Attempting to open a gRPC connection with: "unix://C:\\\\csi\\\\csi.sock"
2023-08-09T14:37:47.687116600+01:00 I0809 14:37:47.686514    1700 connection.go:154] Connecting to unix://C:\\csi\\csi.sock
I0809 14:37:55.779604    1700 main.go:199] Calling CSI driver to discover driver name
2023-08-09T14:37:55.780553500+01:00 I0809 14:37:55.779604    1700 connection.go:183] GRPC call: /csi.v1.Identity/GetPluginInfo
2023-08-09T14:37:55.783858400+01:00 I0809 14:37:55.779604    1700 connection.go:184] GRPC request: {}
2023-08-09T14:37:55.790137000+01:00 I0809 14:37:55.789639    1700 connection.go:186] GRPC response: {"name":"csi.vsphere.vmware.com","vendor_version":"v3.0.1"}
2023-08-09T14:37:55.790199000+01:00 I0809 14:37:55.789830    1700 connection.go:187] GRPC error: <nil>
2023-08-09T14:37:55.790281100+01:00 I0809 14:37:55.789938    1700 main.go:209] CSI driver name: "csi.vsphere.vmware.com"
2023-08-09T14:37:55.791613600+01:00 I0809 14:37:55.791284    1700 node_register.go:53] Starting Registration Server at: /registration/csi.vsphere.vmware.com-reg.sock
2023-08-09T14:37:55.793067100+01:00 I0809 14:37:55.792431    1700 node_register.go:62] Registration Server started at: /registration/csi.vsphere.vmware.com-reg.sock
2023-08-09T14:37:55.793735200+01:00 I0809 14:37:55.793351    1700 node_register.go:92] Skipping HTTP server because endpoint is set to: ""
2023-08-09T14:37:57.468773300+01:00 I0809 14:37:57.468046    1700 main.go:102] Received GetInfo call: &InfoRequest{}
2023-08-09T14:37:57.471666700+01:00 I0809 14:37:57.471318    1700 main.go:109] "Kubelet registration probe created" path="C:\\var\\lib\\kubelet\\plugins\\csi.vsphere.vmware.com\\registration"
I0809 14:37:58.458318    1700 main.go:121] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}
```

## Engineering Testing
### Manual Testing
After troubleshooting and identifying the cause of the issue, I was able to make the change to the environment variable and see that the pod successfully registered the windows node and I was then able to mount PVC's to windows pods without issue. I have since applied this change multiple times to different environment's provisioned with Terraform with the same successful results. 

### Automated Testing


## QA Testing Considerations


### Regressions Considerations


## Backporting considerations
In my opinion this change should be backported as the current state of the rancher-vsphere-csi chart does not work on Windows as documentation suggests. I was not able to get this to work until I figured out the root cause of the issue, nor was I able to find any documentation on how to fix this. If the node-driver-registrarr pod ran in a powershell environment, the path might work fine, but it currently does not.
